Fix calling dtor on an abstract class with non-virtual dtor

davidlt opened the pull request:
Clang pre-3.9 fails with an error:

```
In file included from src/TrackingTools/GeomPropagators/src/Aq:12:
TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h:42:34: error: destructor called on 'HelixPlaneCrossing' that is abstract but has non-virtual destructor [-Werror,-Wdelete-non-virtual-dtor]
  ~OptimalHelixPlaneCrossing() { get()->~Base(); }
                                 ^
src/TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h:42:41: note: qualify call to silence this warning
  ~OptimalHelixPlaneCrossing() { get()->~Base(); }
                                        ^
                                        HelixPlaneCrossing::
1 error generated.
```

The changeset in Clang: http://reviews.llvm.org/D16206

Quote from C++ Standard used in this Clang changeset:

```
C++ [expr.delete]p3:
  In the first alternative (delete object), if the static type of the
  object to be deleted is different from its dynamic type, the static
  type shall be a base class of the dynamic type of the object to be
  deleted and the static type shall have a virtual destructor or the
  behavior is undefined.
```

Signed-off-by: David Abdurachmanov David.Abdurachmanov@cern.ch

cmsbuild commented:
A new Pull Request was created by  @davidlt  for CMSSW_8_1_X.

It involves the following packages:

TrackingTools/GeomPropagators


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@makortel, @GiacomoSguazzoni, @rovere, @VinInn, @bellan, @gpetruc, @dgulhan this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

davidlt commented:
DEVEL IBs contain LLVM/Clang/LDD/RT/etc built from master branch to work on CMSSW + C++ Modules.

So, to avoid undefined behaviour one can also probably add virtual dtor to the class. Up to the experts to decide how to resolve this.

VinInn commented:
yes, HelixPlaneCrossing  needs a virtual destructor... (It is pure virtual)
do you mind to add one in this very same PR?

Commit:
Fix calling dtor on an abstract class with non-virtual dtor

Clang pre-3.9 fails with an error:

    In file included from src/TrackingTools/GeomPropagators/src/Aq:12:
    TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h:42:34: error: destructor called on 'HelixPlaneCrossing' that is abstract but has non-virtual destructor [-Werror,-Wdelete-non-virtual-dtor]
      ~OptimalHelixPlaneCrossing() { get()->~Base(); }
                                     ^
    src/TrackingTools/GeomPropagators/interface/OptimalHelixPlaneCrossing.h:42:41: note: qualify call to silence this warning
      ~OptimalHelixPlaneCrossing() { get()->~Base(); }
                                            ^
                                            HelixPlaneCrossing::
    1 error generated.

The changeset in Clang: http://reviews.llvm.org/D16206

Quote from C++ Standard used in this Clang changeset:

    C++ [expr.delete]p3:
      In the first alternative (delete object), if the static type of the
      object to be deleted is different from its dynamic type, the static
      type shall be a base class of the dynamic type of the object to be
      deleted and the static type shall have a virtual destructor or the
      behavior is undefined.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

cmsbuild commented:
Pull request #15148 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

davidlt commented:
Updated with virtual dtor.

VinInn commented:
@cmsbuild , please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #15148 was updated. @dmitrijus, @cvuosalo, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 0fa9ffd7629cfaf085f4c1b08447df4d91009bf7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15148/13945/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15148/13945/summary.html

smuzaffar commented:
sorry, in order to get the IB pages recognize latest IBs I had to amend the e495908 commit to have message like "Merge CMSSW_8_0_X into CMSSW_8_1_X.". Due to that now to old commit is shows as part of this PR. either ignore it (it is already part of IB) or please rebase based on CMSSW_8_1_X head.

you can start from CMSSW_8_1_X and cherry pick the commits.

cmsbuild commented:
Pull request #15148 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 6319f5dfda1e0ad33ed931f8e5d76e926a7b3c50
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15148/13954/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15148/13954/summary.html

cvuosalo commented:
+1

For #15148 6319f5dfda1e0ad33ed931f8e5d76e926a7b3c50

Adding a virtual destructor to HelixPlaneCrossing to silence a CLANG error. There should be no change in run-time behavior.

The code fix is satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-07-08-1700 show no significant differences, as expected.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

davidlange6 commented:
+1
